In [1]:
import pandas as pd

In [2]:
import os
print(os.getcwd())

C:\Users\Amalitech\Downloads\week5_assignment


In [3]:
import os
print(os.listdir())

['.ipynb_checkpoints', 'part1.3&2.png', 'part3&4.png', 'part3.png', 'part4.3(a)&4.3(c).png', 'part5.png', 'Screenshot 2026-10-03 221106.png', 'week5_kigali_weather.csv', 'week5_Nishimwe_prosper_28926.ipynb', 'weet days&reflection_part6.png']


In [4]:
df = pd.read_csv("week5_kigali_weather.csv")

In [5]:
df.head()

,day,date,temp_c,rainfall_mm,humidity_pct
0,1,01/09/2026,24.4,0.0,77.0
1,2,02/09/2026,25.9,0.0,74.7
2,3,03/09/2026,24.5,0.0,72.1
3,4,04/09/2026,26.0,0.0,76.3
4,5,05/09/2026,24.6,22.9,73.2


In [6]:
import numpy as np

dates = df["date"].values
temps = df["temp_c"].values
rain = df["rainfall_mm"].values
hum = df["humidity_pct"].values

In [7]:
fahrenheit = [(temp * 9/5) + 32 for temp in temps]

fahrenheit[:5]

[np.float64(75.92),
 np.float64(78.62),
 np.float64(76.1),
 np.float64(78.8),
 np.float64(76.28)]

In [8]:
hot_days = [temp for temp in temps if temp >= 26]

hot_days

[np.float64(26.0),
 np.float64(26.1),
 np.float64(27.6),
 np.float64(26.0),
 np.float64(27.4),
 np.float64(27.2),
 np.float64(28.5),
 np.float64(26.8),
 np.float64(28.1),
 np.float64(26.3),
 np.float64(27.5),
 np.float64(28.6),
 np.float64(26.7),
 np.float64(27.8),
 np.float64(26.8),
 np.float64(27.8),
 np.float64(26.6),
 np.float64(26.2)]

In [9]:
cool_days = [round(temp, 1) for temp in temps if temp < 24]

cool_days

[np.float64(22.7),
 np.float64(23.6),
 np.float64(22.3),
 np.float64(23.1),
 np.float64(21.0),
 np.float64(21.9),
 np.float64(22.8),
 np.float64(20.8),
 np.float64(21.7),
 np.float64(19.7),
 np.float64(20.8),
 np.float64(21.8),
 np.float64(20.0),
 np.float64(21.1),
 np.float64(19.3),
 np.float64(20.5),
 np.float64(21.8),
 np.float64(20.1),
 np.float64(21.4),
 np.float64(19.8),
 np.float64(21.2),
 np.float64(22.7),
 np.float64(21.1),
 np.float64(22.6),
 np.float64(21.1),
 np.float64(22.7),
 np.float64(22.8),
 np.float64(22.9)]

In [10]:
raw = ["24.4", "25.9", "28.6", "19.3"]

In [11]:
raw_floats = [float(value) for value in raw]

print(raw_floats)
print(sum(raw_floats) / len(raw_floats))

[24.4, 25.9, 28.6, 19.3]
24.55


In [12]:
# Exercise 1.1
fahrenheit = [(temp * 9/5) + 32 for temp in temps]
print(fahrenheit[:5])

[np.float64(75.92), np.float64(78.62), np.float64(76.1), np.float64(78.8), np.float64(76.28)]


In [13]:
# Exercise 1.2(a)
hot_days = [temp for temp in temps if temp >= 26]
print(hot_days)

[np.float64(26.0), np.float64(26.1), np.float64(27.6), np.float64(26.0), np.float64(27.4), np.float64(27.2), np.float64(28.5), np.float64(26.8), np.float64(28.1), np.float64(26.3), np.float64(27.5), np.float64(28.6), np.float64(26.7), np.float64(27.8), np.float64(26.8), np.float64(27.8), np.float64(26.6), np.float64(26.2)]


In [14]:
# Exercise 1.2(b)
cool_days = [round(temp) for temp in temps if temp < 24]
print(cool_days)

[23, 24, 22, 23, 21, 22, 23, 21, 22, 20, 21, 22, 20, 21, 19, 20, 22, 20, 21, 20, 21, 23, 21, 23, 21, 23, 23, 23]


In [15]:
# Exercise 1.3
raw = ["24.4", "25.9", "28.6", "19.3"]

raw_floats = [float(value) for value in raw]

print(raw_floats)
print(sum(raw_floats) / len(raw_floats))

[24.4, 25.9, 28.6, 19.3]
24.55


In [16]:
cool_days = [round(temp) for temp in temps if temp < 24]

In [17]:
results = [
    ("Aline", 85),
    ("Eric", 78),
    ("Diane", 91),
    ("Samuel", 92),
    ("Grace", 76)
]

ranked = sorted(results, key=lambda x: x[1], reverse=True)

print(ranked[:3])

[('Samuel', 92), ('Diane', 91), ('Aline', 85)]


In [18]:
rain_log = [(3, 0.0), (7, 12.5), (11, 3.2), (14, 26.2), (18, 8.9)]

wettest_day = max(rain_log, key=lambda x: x[1])

print(wettest_day)

(14, 26.2)


part 3

In [19]:
import time

arr = np.arange(1_000_000)

start = time.time()
loop_result = [x * 2 for x in arr]
loop_time = time.time() - start

start = time.time()
numpy_result = arr * 2
numpy_time = time.time() - start

print(f"Python loop time: {loop_time:.6f} seconds")
print(f"NumPy time: {numpy_time:.6f} seconds")
print(f"NumPy was approximately {loop_time / numpy_time:.1f} times faster")

Python loop time: 0.170381 seconds
NumPy time: 0.003033 seconds
NumPy was approximately 56.2 times faster


In [20]:
print(f"Mean temperature: {temps.mean():.2f} °C")
print(f"Maximum temperature: {temps.max():.2f} °C")
print(f"Minimum temperature: {temps.min():.2f} °C")
print(f"Standard deviation: {temps.std():.2f} °C")

Mean temperature: 24.00 °C
Maximum temperature: 28.60 °C
Minimum temperature: 19.30 °C
Standard deviation: 2.61 °C


In [21]:
position = np.argmax(temps)

print(f"Hottest day: {dates[position]}")
print(f"Temperature: {temps[position]} °C")

Hottest day: 17/09/2026
Temperature: 28.6 °C


Part 4 — Boolean Masking

In [22]:
hot_mask = temps >= 26

print(f"Number of hot days: {hot_mask.sum()}")
print(f"Average temperature on hot days: {temps[hot_mask].mean():.2f} °C")

Number of hot days: 18
Average temperature on hot days: 27.11 °C


In [23]:
rain_mask = rain > 0

print(f"Number of rainy days: {rain_mask.sum()}")
print(f"Total rainfall: {rain.sum():.1f} mm")
print(f"Heaviest single-day rainfall: {rain.max():.1f} mm")

Number of rainy days: 16
Total rainfall: 331.2 mm
Heaviest single-day rainfall: 26.2 mm


Exercise 4.3(a) — Rainy vs dry

In [24]:
rain_mask = rain > 0

print(f"Average temperature on rainy days: {temps[rain_mask].mean():.2f} °C")
print(f"Average temperature on dry days: {temps[~rain_mask].mean():.2f} °C")

Average temperature on rainy days: 24.14 °C
Average temperature on dry days: 23.95 °C


Exercise 4.3(b) — Hot AND rainy

In [25]:
hot_and_rainy = (temps >= 26) & (rain > 0)

print(f"Hot and rainy days: {hot_and_rainy.sum()}")

Hot and rainy days: 5


Exercise 4.3(c) — Humidity ≥ 70%

In [26]:
humidity_mask = hum >= 70

print(f"Days with humidity >= 70%: {humidity_mask.sum()}")

Days with humidity >= 70%: 16


Part 5 — Anomalies & Trends

In [27]:
anomaly = temps - temps.mean()

above_mask = anomaly > 3

print("Dates more than 3 °C above the mean:")
print(dates[above_mask])

below_mask = anomaly < -3

print(f"Days more than 3 °C below the mean: {below_mask.sum()}")

Dates more than 3 °C above the mean:
<ArrowStringArray>
['07/09/2026', '09/09/2026', '11/09/2026', '12/09/2026', '14/09/2026',
 '16/09/2026', '17/09/2026', '19/09/2026', '22/09/2026']
Length: 9, dtype: str
Days more than 3 °C below the mean: 8


In [28]:
first_30 = temps[:30]
last_30 = temps[30:]

print(f"Average temperature, first 30 days: {first_30.mean():.2f} °C")
print(f"Average temperature, last 30 days: {last_30.mean():.2f} °C")

# Kigali got cooler because the average temperature decreased.

Average temperature, first 30 days: 26.21 °C
Average temperature, last 30 days: 21.78 °C


Top 5 wettest days

In [29]:
wettest_days = sorted(
    zip(dates, rain),
    key=lambda x: x[1],
    reverse=True
)

for date, rainfall in wettest_days[:5]:
    print(f"{date}: {rainfall} mm")

07/09/2026: 26.2 mm
14/09/2026: 25.1 mm
21/09/2026: 24.0 mm
05/09/2026: 22.9 mm
28/09/2026: 22.9 mm


### Reflection

**1. What did NumPy hide from you , and why was it still important to learn the loop version first?**

NumPy hid the repetitive process of going through each value one by one. Instead of writing a loop to calculate an average, I could simply use `temps.mean()`. Learning loops first was important because it helped me understand the basic logic behind the calculation. NumPy then made the same operations shorter, faster, and easier to apply to large datasets.

**2. One insight about Kigali's weather**

The analysis revealed a clear cooling trend during the 60 days. The average temperature during the first 30 days was about 26.21 °C, while the average during the last 30 days was about 21.78 °C. The hottest day was 17/09/2026, with a temperature of 28.6 °C.